File: micro-service-hello-client/src/main/java/demo/HelloClientApplication.java
package demo;

import feign.*;
import org.springframework.boot.*;
import org.springframework.boot.autoconfigure.*;
import org.springframework.cloud.client.discovery.*;
import org.springframework.cloud.openfeign.*;
import org.springframework.context.annotation.*;
import org.springframework.web.bind.annotation.*;

/**
 * @author Chen
 */
@SpringBootApplication
@EnableDiscoveryClient
@EnableFeignClients
public class HelloClientApplication {

  public static void main(String[] args) {
    SpringApplication.run(HelloClientApplication.class, args);
  }


  @FeignClient("HELLO-SERVER")
  interface HelloClient {

    @ResponseBody
    @RequestMapping(value = "/get", method = RequestMethod.GET)
    User hello1(@RequestParam("name") String name);

    @ResponseBody
    @RequestMapping(value = "/post", method = RequestMethod.POST)
    User hello2(@RequestParam("name") String name);

  }

  @Bean
  public Logger.Level feignLoggerLevel() {
    return feign.Logger.Level.FULL;
  }

}

File: micro-service-hello-client/src/main/java/demo/HelloClientApplication.kt
package demo

import feign.Logger
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.boot.task.ThreadPoolTaskSchedulerCustomizer
import org.springframework.cloud.client.discovery.EnableDiscoveryClient
import org.springframework.cloud.openfeign.EnableFeignClients
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import java.util.concurrent.*


fun main(args: Array<String>) {
  runApplication<HelloClientApplication>(*args)
}


@SpringBootApplication
@EnableDiscoveryClient
@EnableFeignClients
class HelloClientApplication {

  @Bean
  fun feignLoggerLevel(): Logger.Level {
    return Logger.Level.FULL
  }

}


@Configuration
class ScheduledGracefulShutdownConfig {
  //this shouldn't be necessary at all with ' spring.task.scheduling.shutdown.await-termination'. but neither works
  @Bean
  fun taskSchedulerCustomizer(): ThreadPoolTaskSchedulerCustomizer {
    return ThreadPoolTaskSchedulerCustomizer { taskScheduler ->
      taskScheduler.setWaitForTasksToCompleteOnShutdown(true)
      taskScheduler.setAwaitTerminationSeconds(60)
    }
  }
}

@Service
class ScheduledService {
  private val LOGGER = LoggerFactory.getLogger(ScheduledService::class.java)

  @Scheduled(fixedRate = 100000000L)
  @Throws(InterruptedException::class)
  fun scheduled() {
    LOGGER.info("Starting scheduled job...")
    TimeUnit.MINUTES.sleep(5)
    LOGGER.info("Scheduled job finished.")
  }
}
